addons/mass_mailing/wizard/mail_compose_message.py

```diff
@@ -20,60 +20,83 @@ def get_mail_values(self, res_ids):
+                                                       {'body': mail_values['body_html']},
+                                                       minimal_qcontext=True, raise_if_not_found=False)
                     if body:
                         mail_values['body_html'] = body
 
-                trace_vals = {
+                traces_vals = [{
                     'model': self.model,
                     'res_id': res_id,
                     'mass_mailing_id': mass_mailing.id,
-                    # if mail_to is void, keep falsy values to allow searching / debugging traces
-                    'email': recipients_info[res_id]['mail_to'][0] if recipients_info[res_id]['mail_to'] else '',
-                }
+                    'email': email,
+                } for email in (recipients_info[res_id]['mail_to'] or [''])]  # if mail_to is void, keep falsy values to allow searching / debugging traces
+
                 # propagate failed states to trace when still-born
-                if mail_values.get('state') == 'cancel':
-                    trace_vals['trace_status'] = 'cancel'
-                elif mail_values.get('state') == 'exception':
-                    trace_vals['trace_status'] = 'error'
-                if mail_values.get('failure_type'):
-                    trace_vals['failure_type'] = mail_values['failure_type']
+                for trace_vals in traces_vals:
+                    if mail_values.get('state') == 'cancel':
+                        trace_vals['trace_status'] = 'cancel'
+                    elif mail_values.get('state') == 'exception':
+                        trace_vals['trace_status'] = 'error'
+                    if mail_values.get('failure_type'):
+                        trace_vals['failure_type'] = mail_values['failure_type']
 
                 mail_values.update({
                     'mailing_id': mass_mailing.id,
-                    'mailing_trace_ids': [(0, 0, trace_vals)],
+                    'mailing_trace_ids': [(0, 0, trace_vals) for trace_vals in traces_vals],
                     # email-mode: keep original message for routing
                     'is_notification': mass_mailing.reply_to_mode == 'update',
-                    'auto_delete': not mass_mailing.keep_archives,
                 })
+                # we need to keep mails to link messages to traces with archives
+                # we do not need to keep mails if we do not want archives
+                mail_values['auto_delete'] = not mass_mailing.keep_archives
+
         return res
 
+    def _get_mail_values_auto_delete(self):
+        if self.mass_mailing_id.keep_archives:
+            return False
+        return super()._get_mail_values_auto_delete()
+
+    def _should_add_traces(self):
+        return self.composition_mode == 'mass_mail' and \
+            (self.mass_mailing_name or self.mass_mailing_id) and \
+            self.env['ir.model'].sudo().search_count([('model', '=', self.model), ('is_mail_thread', '=', True)])
```

Comment:
a bit longer, but it's better to use `()` when possible instead of `\`
```suggestion
        return (
            self.composition_mode == 'mass_mail'
            and (self.mass_mailing_name or self.mass_mailing_id)
            and self.env['ir.model'].sudo().search_count(
                [('model', '=', self.model), ('is_mail_thread', '=', True)])
        )
```